Migrate SpectrumAnalysisIACT to the high-level interface

Bultako opened the pull request:
This PR aims to add all features needed in the high-level interface to perform a 1D spectrum analysis. 

```
In [1]: import yaml                                                                                                                                                                   

In [2]: from gammapy.scripts import Analysis                                                                                                                                          

In [3]: config = """
model:
    components:
        - name: Crab
          spectral:
            type: PowerLaw
            parameters:
            - name: index
              value: 2.6
              factor: 2.2
              scale: 1.0
              unit: ''
              min: .nan
              max: .nan
              frozen: false
            - name: amplitude
              value: 5.0e-11
              factor: 2.3
              scale: 1.0e-12
              unit: cm-2 s-1 TeV-1
              min: .nan
              max: .nan
              frozen: false
            - name: reference
              value: 1.0
              factor: 1.0
              scale: 1.0
              unit: TeV
              min: .nan
              max: .nan
              frozen: true
flux:
    fp_binning:
        lo_bnd: 1
        hi_bnd: 10
        nbin: 11
        unit: TeV
        interp: log
"""                                                                                                                                                                       

In [4]: config = yaml.load(config)                                                                                                                                                        
/Users/jer/anaconda/envs/gammapy-dev/bin/ipython:1: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  #!/Users/jer/anaconda/envs/gammapy-dev/bin/python

In [5]: a = Analysis(config)   
INFO:gammapy.scripts.analysis:Setting logging parameters (INFO).

In [6]: print(a.config)                                                                                                                                                               
fit: {}
flux:
  fp_binning:
    hi_bnd: 10
    interp: log
    lo_bnd: 1
    nbin: 11
    unit: TeV
general:
  logging:
    level: INFO
  outdir: .
geometry:
  axes: {}
model:
  components:
  - name: Crab
    spectral:
      parameters:
      - factor: 2.2
        frozen: false
        max: .nan
        min: .nan
        name: index
        scale: 1.0
        unit: ''
        value: 2.6
      - factor: 2.3
        frozen: false
        max: .nan
        min: .nan
        name: amplitude
        scale: 1.0e-12
        unit: cm-2 s-1 TeV-1
        value: 5.0e-11
      - factor: 1.0
        frozen: true
        max: .nan
        min: .nan
        name: reference
        scale: 1.0
        unit: TeV
        value: 1.0
      type: PowerLaw
observations:
  datastore: $GAMMAPY_DATA/hess-dl3-dr1
  filters:
  - filter_type: par_value
    value_param: Crab
    variable: TARGET_NAME
reduction:
  background:
    background_estimator: reflected
    on_region:
      center:
      - 83.633 deg
      - 22.014 deg
      frame: icrs
      radius: 0.1 deg
  containment_correction: true
  data_reducer: 1d


In [7]: a.get_observations()                                                                                                                                                          
INFO:gammapy.scripts.analysis:Fetching observations.
INFO:gammapy.scripts.analysis:Info for OBS_ID = 23592
- Start time: 53347.91
- Pointing pos: RA 82.01 deg / Dec 22.01 deg
- Observation duration: 1686.0 s
- Dead-time fraction: 6.212 %

INFO:gammapy.scripts.analysis:Info for OBS_ID = 23523
- Start time: 53343.92
- Pointing pos: RA 83.63 deg / Dec 21.51 deg
- Observation duration: 1687.0 s
- Dead-time fraction: 6.240 %

INFO:gammapy.scripts.analysis:Info for OBS_ID = 23526
- Start time: 53343.95
- Pointing pos: RA 83.63 deg / Dec 22.51 deg
- Observation duration: 1683.0 s
- Dead-time fraction: 6.555 %

INFO:gammapy.scripts.analysis:Info for OBS_ID = 23559
- Start time: 53345.96
- Pointing pos: RA 85.25 deg / Dec 22.01 deg
- Observation duration: 1686.0 s
- Dead-time fraction: 6.398 %


In [8]: a.reduce()                                                                                                                                                                    
INFO:gammapy.scripts.analysis:Reducing data sets.
INFO:gammapy.spectrum.reflected:Found 46 reflected regions for the Obs #23592
INFO:gammapy.spectrum.reflected:Found 14 reflected regions for the Obs #23523
INFO:gammapy.spectrum.reflected:Found 14 reflected regions for the Obs #23526
INFO:gammapy.spectrum.reflected:Found 46 reflected regions for the Obs #23559
INFO:gammapy.spectrum.extract:Running <gammapy.spectrum.extract.SpectrumExtraction object at 0x10332bcc0>
INFO:gammapy.spectrum.extract:Process observation
 Info for OBS_ID = 23592
- Start time: 53347.91
- Pointing pos: RA 82.01 deg / Dec 22.01 deg
- Observation duration: 1686.0 s
- Dead-time fraction: 6.212 %

INFO:gammapy.spectrum.extract:Update observation meta info
INFO:gammapy.spectrum.extract:Offset : 1.501568530069493 deg

INFO:gammapy.spectrum.extract:Fill events
INFO:gammapy.spectrum.extract:Extract IRFs
INFO:gammapy.spectrum.extract:Apply containment correction
/Users/jer/git/gammapy/gammapy/gammapy/spectrum/extract.py:232: RuntimeWarning: invalid value encountered in true_divide
  self.containment = new_aeff.data.data.value / self._aeff.data.data.value
INFO:gammapy.spectrum.extract:Process observation
 Info for OBS_ID = 23523
- Start time: 53343.92
- Pointing pos: RA 83.63 deg / Dec 21.51 deg
- Observation duration: 1687.0 s
- Dead-time fraction: 6.240 %

INFO:gammapy.spectrum.extract:Update observation meta info
INFO:gammapy.spectrum.extract:Offset : 0.4995557435558715 deg

INFO:gammapy.spectrum.extract:Fill events
INFO:gammapy.spectrum.extract:Extract IRFs
INFO:gammapy.spectrum.extract:Apply containment correction
INFO:gammapy.spectrum.extract:Process observation
 Info for OBS_ID = 23526
- Start time: 53343.95
- Pointing pos: RA 83.63 deg / Dec 22.51 deg
- Observation duration: 1683.0 s
- Dead-time fraction: 6.555 %

INFO:gammapy.spectrum.extract:Update observation meta info
INFO:gammapy.spectrum.extract:Offset : 0.5004444451150709 deg

INFO:gammapy.spectrum.extract:Fill events
INFO:gammapy.spectrum.extract:Extract IRFs
INFO:gammapy.spectrum.extract:Apply containment correction
INFO:gammapy.spectrum.extract:Process observation
 Info for OBS_ID = 23559
- Start time: 53345.96
- Pointing pos: RA 85.25 deg / Dec 22.01 deg
- Observation duration: 1686.0 s
- Dead-time fraction: 6.398 %

INFO:gammapy.spectrum.extract:Update observation meta info
INFO:gammapy.spectrum.extract:Offset : 1.5021898826765052 deg

INFO:gammapy.spectrum.extract:Fill events
INFO:gammapy.spectrum.extract:Extract IRFs
INFO:gammapy.spectrum.extract:Apply containment correction

In [9]: a.fit()                                                                                                                                                                       
INFO:gammapy.scripts.analysis:Reading model.
INFO:gammapy.scripts.analysis:PowerLaw

Parameters: 

	   name     value   error      unit      min max frozen
	--------- --------- ----- -------------- --- --- ------
	    index 2.600e+00   nan                nan nan  False
	amplitude 5.000e-11   nan cm-2 s-1 TeV-1 nan nan  False
	reference 1.000e+00   nan            TeV nan nan   True
INFO:gammapy.scripts.analysis:Fitting data sets to model.
INFO:gammapy.scripts.analysis:OptimizeResult

	backend    : minuit
	method     : minuit
	success    : True
	message    : Optimization terminated successfully.
	nfev       : 36
	total stat : 130.56


In [10]: a.get_flux_points()                                                                                                                                                           
INFO:gammapy.scripts.analysis:Calculating flux points.
INFO:gammapy.scripts.analysis:
      e_ref               ref_flux                 dnde                 dnde_ul                dnde_err        is_ul
       TeV              1 / (cm2 s)          1 / (cm2 s TeV)        1 / (cm2 s TeV)        1 / (cm2 s TeV)          
------------------ ---------------------- ---------------------- ---------------------- ---------------------- -----
1.1364636663857248  9.449329992294733e-12  2.816977541295253e-11  3.573691483792928e-11  3.505225434375799e-12 False
1.3768571648527583 3.4113700786811144e-12  1.828679888324125e-11  2.603817913822835e-11 3.4450407931548733e-12 False
1.6681005372000581  4.982516507677505e-12  1.365332939913724e-11  1.755910853352502e-11  1.799039326424294e-12 False
2.1544346900318834 3.2519752617892428e-12 6.1484585038692114e-12  8.309695131203611e-12   9.78342227704282e-13 False
2.6101572156825363 1.1740188049020575e-12  3.861118980517589e-12  6.052405645173305e-12  9.421227414399863e-13 False
3.1622776601683777 1.7147269105466929e-12 2.5897875508619773e-12  3.649825162635732e-12  4.731485364864076e-13 False
 3.831186849557287   6.19045803302273e-13  1.156618808270176e-12 2.1238966040340406e-12  3.955211988710733e-13 False
4.6415888336127775  9.041545956088386e-13  7.045665488041625e-13 1.1577572591321411e-12 1.9297021903699566e-13 False
 5.994842503189405  5.901211512741317e-13 3.6988307657563386e-13  6.485911472662804e-13 1.1511026416367203e-13 False
  7.26291750173621 2.1304384965865337e-13  4.204034950709673e-13  8.073308222764743e-13 1.5501752336572198e-13 False
  8.79922543569107 3.1116368887006885e-13 1.1711100306620158e-13  2.544714864537633e-13  5.304630090176535e-14 False

In [11]: a.flux_points_dataset.peek()                                                                                                                                                 
Out[11]: 
(<matplotlib.axes._subplots.AxesSubplot at 0x1111aec18>,
 <matplotlib.axes._subplots.AxesSubplot at 0x111b709b0>)
```

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -51,6 +60,124 @@ def settings(self):
+        self.config.validate()
+
+        # create geometry
+        self.geom = WcsGeom.create(
```
I think data reduction has to take either the 1D or 3D path, i.e. decide and have different methods for those.
For 1D no WCS object is needed, that should be in the 3D path (and probably not in this PR)?

Bultako replied:
umm.. in principle I agree, but I think I need the geometry to create an eventual exclusion mask that will be (un-)considered in the reflected regions methof to estimate the background. is it possible to create an exclusion mask without a geometry?

cdeil replied:
Exclusion mask is difficult. You'd also have to specify config which regions to exclude.
Suggest to take an exclusion mask FITS filename, optionally.
Whether to later also allow config to create exclusion regions I'm not sure.

Of course, for 3D will will need to create a geom from config, so don't throw the code away.
I don't know yet whether we want to limit the use of config to Analysis, or use it throughout Gammapy, i.e. in this case there could be a MapGeom.from_config(config_dict) that takes this part of the config. Maybe develop that in utility functions in analysis.py for now?
I probably won't be able to join the dev call tomorrow, this is something you should also bring up as a question and gather more opinions there.

Bultako replied:
Ok, I've put it the *build geometry* part aside as a helper function and deal with exclusion_mask as a FITS file.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -51,6 +60,124 @@ def settings(self):
+            log.error("Datastore {} not found.".format(datastore_path))
+            return False
+
+        ids = set()
```
This reformatting of obs select config is super long. Can it be avoided by directly using the existing
https://docs.gammapy.org/0.13/api/gammapy.data.ObservationTable.html#gammapy.data.ObservationTable.select_observations
or by modifying that method or adding a new one on obs-table (maybe with a TODO comment or reminder issue to unify the two methods again in the future)?

Note that Analysis should only dispatch to existing functionality throughout Gammapy, it shouldn't re-implement new functionality. The code you're adding would need review and work, if really a lot of code is needed obs selection API & implementation & tests should be a separate PR ideally.

Bultako replied:
Yes, I agree this all or most of this should be handled at `ObservationTable.select_observations`. I have reduced a little bit the amount of code, I will create a PR to improve `ObservationTable.select_observations` and get rid of all these tens of lines.

cdeil replied:
Thanks!

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -51,6 +60,124 @@ def settings(self):
+            self.datastore = DataStore().from_dir(datastore_path)
+        else:
+            log.error("Datastore {} not found.".format(datastore_path))
+            return False
```
Shouldn't we raise an error?
It's very easy to forget to look at the return value of this method and have a silent error.
Note that you don't even document the return value and how callers should handle it.
If you just raise an error, I think that's fine, no extra documentation needed.

Bultako replied:
Ok.
I raise now a *`FileNotFoundError`*

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -32,14 +39,16 @@ class Analysis:
     def __init__(self, config=None):
         self._config = Config(config)
         self._set_logging()
+        self.datastore = DataStore()
```
Suggest to init to None. I don't think having an empty DataStore or Observations object can help in any way, but it has the potential to have silent errors when users run and nothing happens because they have zero runs or observations.

I'm not sure what API and responsibilities we want concerning how Analysis interacts with DataStore and Observations. My guess would be that Analysis should have the observations data member or property (created from config), but doesn't need the `self.datastore`. The DataStore is just something that's used at the start to create the `Observations` object, but then never needed during analysis?

This Analysis class will quickly grow and become very complex, because it's the manager for all the analysis workflows. So we have to work hard to keep it as simple and small as possible, to keep it maintainable.

Bultako replied:
Ok.
I've moved to `__init__` the declaration of all the instance *has-a*  properties defined by composition of other classes (datastore is not among them). I think it's good to have them listed at the beginning. 

```
        self.observations = None
        self.geom = None
        self.background_estimator = None
        self.extraction = None
```

cdeil replied:
OK, let's see how it goes ...

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -51,6 +60,124 @@ def settings(self):
+
+    def reduce_data(self):
+        """Produce reduced data sets."""
+        self.config.validate()
```
Remove this `self.config.validate` call in `reduce_data`?
Generic config validation should happen in Analysis init.
If specific validation is needed concerning config and data reduction (probably not?), there should be an extra method for that.

Bultako replied:
I think we need a validation of some kind before most/all of the actions of the high-level interface (i.e. get_observations, reduce_data, optimize, etc.) In principle, the user may add/modif config settings in between these actions, so a validation of the settings is desirable. For the moment I use the generic one because until now I manage to put all the conditions in the schema, but if we consider that more complex specific validation is needed I agree this should be in an extra method.

Bultako replied:
Well, we do have a *__validate_reduction_settings* helper function now. I needed that.

cdeil commented on gammapy/scripts/tests/test_analysis.py:
```diff
@@ -21,3 +22,79 @@ def test_validate_astropy_quantities():
+            "result": 2,
+        },
+    ],
+)
```
I wonder if we should start using YAML multi-line text in Gammapy tests instead of dicts, because it's a bit easier to read and write and shorter?
```
    config = """
    observations:
        datastore: ...
        filter: ...
    """
```
@Bultako - Better or worse than the current dicts?

Bultako replied:
Yes! I take it.
Do you have an example on how to do it?

cdeil replied:
Can you try to copy & paste your dict from the test Python file, and use `yaml.dump` to get the corresponding string? Like I said, I think this will be more readable and shorter for config, but I'm not 100% sure.

Bultako replied:
Ok, I've done this.
I think it is more readable now.
https://github.com/gammapy/gammapy/blob/d26e574cb7a32469032cca74fc9a9f3894b19ec1/gammapy/scripts/tests/test_analysis.py#L28-L86

cdeil replied:
Agreed, it's better.

I think we'll want a mini language, e.g. for regions we use DS9 strings, and for other things we use strings like "zenith > 10" and parse them.

But this really isn't high-priority, let's leave as-is and come back to obs selection later.

cdeil commented on gammapy/scripts/config/schema.yaml:
```diff
@@ -44,47 +44,161 @@ properties:
+                        frame:
+                            default: icrs
+                            enum: [galactic, equatorial, icrs, fk5]
+                        lon: {type: number, default: 83.633 deg}
```
If the default is "83.633 deg" the type has to be string or quantity, no?
You write `type: number`.
!?

There's a few config items where probably we don't want to offer a default, target position being one of them.

Otherwise people will forget to change it, and run a Crab position analysis, which will not be what they wanted in 99% of the cases.

Is it possible to have config items without a default in our config scheme?
Can you make the target position required?

Bultako replied:
*re: type: number*
we have our own validator that extends the one of `jsonschema` and re-defines validation conditions (adding support for astropy quantities) to one type among a set of pre-defined types. I chose number instead of string because it was easier and also because quantities have numbers so we inherit the built-in validation for numbers. This is only unsed internally for validation, users do not have to know this subtleties in the jsonschema syntax of `schema.yaml` file, they don't have to provide types nor worrying about if what they provide is a number or a string, as long as it may be converted to an astropy quantity.

Yes, it is possible to have config items without a default in our config scheme. It's up to us to decide what settings have a default, which ones are required and under which conditions. (i.e. the value of a setting may change the default of another or require/disable the presence of others, etc.) 

I am using the `schema.yaml` to have default values for the specific use case I'm facing, we will provide a clean  `schema.yaml`  with good default generic values at the end. FOr the moment, just consider the overall structure of the file, the names of the properties and where they are placed, not the specific values.

cdeil replied:
Would be good to have an example very soon of config items without defaults. There won't be many, and we won't touch them often, because every time you have to adjust most tests and docs examples. Would start with position.

Bultako replied:
Ok, as it was discussed in last dev call, this will be done in another PR.

Bultako commented:
@cdeil 

Ok, to merge this PR soon and avoid having 1000+ lines.
I will add the features to do the fitting and then I think we would have cover `SpectrumAnalysisIACT` use case. I will leave the documentation and notebooks for another PR.

I've addressed some of your suggestions and commented on the rest.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -51,6 +63,118 @@ def settings(self):
+        # """Run all steps for the spectrum extraction."""
+
+        on = self.settings["reduction"]["background"]["on_region"]
+        on_lon = Angle(on["center"][0])
```
Is it possible to move the handling of Quantity, Angle, SkyCoord, Time, SkyRegion into the config file handling in a generic way? I.e. declare those as types in the schema, and then the right object creation is done automatically, either by yaml or jsonschema somewhere?
I think if you use this YAML load (http://docs.astropy.org/en/stable/io/misc.html#example) it will do the trick.
But I don't know how this interplays with json-schema.
What we want is to be able to write strings like "30 deg" or "42 rad" for angles in the config file, and not need to put the `Angle` to type-cast in the Analysis code and elsewehre in Gammapy.
Possible?

Bultako replied:
As it was discussed in last dev call, this will be done in another PR.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -51,6 +63,118 @@ def settings(self):
+            self._spectrum_extraction()
+
+    def _spectrum_extraction(self):
+        # """Run all steps for the spectrum extraction."""
```
Remove hashtag, just put a normal docstring as-is.

cdeil commented on gammapy/scripts/config/schema.yaml:
```diff
@@ -44,47 +44,161 @@ properties:
+    reduction:
+        description: "Process of data reduction."
         type: object
+        default: {}
```
Do we need these `default: {}` at intermediate levels?
I don't think that default empty dict is ever used, because we have defaults for sub-parts.
Remove here and in similar places?

Bultako replied:
We need that when there are nested properties at a deeper level that have default values, it is a way to declare empty data structures.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -51,6 +63,118 @@ def settings(self):
+                    ids.difference_update(selected_obs["OBS_ID"].tolist())
+                else:
+                    ids.update(selected_obs["OBS_ID"].tolist())
+        if len(ids):
```
Can you remove the if here?
`datastore.get_observations` should just be able to handle the empty list case.
If it doesn't already, make it so. But probably it already does.

Bultako commented:
@cdeil 
This PR is ready for final review before merging.
